pkg/controller/workspacerouting/solvers/common.go

```diff
@@ -81,87 +81,105 @@ func getServicesForEndpoints(endpoints map[string][]v1alpha1.Endpoint, workspace
 
 	services = append(services, corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      common.ServiceName(workspaceMeta.WorkspaceId),
-			Namespace: workspaceMeta.Namespace,
+			Name:      common.ServiceName(meta.WorkspaceId),
+			Namespace: meta.Namespace,
 			Labels: map[string]string{
-				config.WorkspaceIDLabel: workspaceMeta.WorkspaceId,
+				config.WorkspaceIDLabel: meta.WorkspaceId,
 			},
 		},
 		Spec: corev1.ServiceSpec{
 			Ports:    servicePorts,
-			Selector: workspaceMeta.PodSelector,
+			Selector: meta.PodSelector,
 			Type:     corev1.ServiceTypeClusterIP,
 		},
 	})
 
 	return services
 }
 
-func getIngressesForSpec(endpoints map[string][]v1alpha1.Endpoint, workspaceMeta WorkspaceMetadata) ([]v1beta1.Ingress, map[string][]v1alpha1.ExposedEndpoint) {
+func getRoutingForSpec(endpoints map[string][]v1alpha1.Endpoint, meta WorkspaceMetadata) ([]v1beta1.Ingress, []routeV1.Route) {
```

Comment:
AFAIU we only need ingresses or routes but not both at the same time. Then I wonder if we're able to build some general interface to work with Routing and that just assign right implementation according to the current platform, like:
```go
struct Routing {
  Name string
  Host string
  Target ServicePortReference
}
interface RoutingManager {
  generateForEndpoint(Endpoint) runtime.Object
  sync([]runtime.Object) error
  resolveURLForEndpoint([]runtime.Object, Endpoint) string, err
}
```

or simpler in the implementation terms ( no needs to cast objects and play with their types)
```go
struct Routing {
  Route routev1.Route
  Ingress ingressv1.Ingress
}
interface RoutingManager {
  generateForEndpoint(Endpoint) Routing
  sync([]Routing) error
  resolveURLForEndpoint([]Routing, Endpoint) string, err
}
```

It would make code simpler in some places but make more difficult in others...
We should consider such refactoring later to play and implement if we like it.

Reply:
Yeah; my primary concern is that this kind of repetition is built into the language. I want to try to avoid shoe-horning OOP-style thinking of "make everything an interface" unless it really increases clarity. 

If we want to start thinking about abstracting out interfaces, the `getClusterX` and `syncX` functions are all practically identical:
```
syncX(workspace, ...) {
  spec := getSpecForX()
  cluster := getClusterForX(name, namespace, client)
  return syncSpecAndCluster(spec, cluster, client)
```
but I'm not ready to fight golang to implement it cleanly. We currently have 5-6 sets of functions where the only difference is the type signature, but abstracting that to an interface means dealing with `runtime.Object` and casting.